src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs

```diff
@@ -97,7 +97,14 @@ protected static ValueSource<TextAndVersion> CreateStrongText(TextLoader loader,
         }
 
         protected static ValueSource<TextAndVersion> CreateRecoverableText(TextAndVersion text, SolutionServices services)
-            => new RecoverableTextAndVersion(CreateStrongText(text), services.TemporaryStorage);
+        {
+            var result = new RecoverableTextAndVersion(CreateStrongText(text), services.TemporaryStorage);
+
+            // Ensure the recoverable text is saved in temporary storage
+            result.GetValue();
+
+            return result;
```

Comment:
the implications of this are not clear to me.  @jasonmalinowski do you know this part any better?

Reply:
so we use memory mapped files for this right?  is there a concern about what this does for using up address space?

Reply:
The call to `GetValue` is just forcing this line to run:

https://github.com/dotnet/roslyn/blob/a37aeb14c5608c3c213a6bc9ce0f9bc0b4c9fa10/src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs#L83

Reply:
sure.  but there are two things that aren't clear to me:

1. how expensive all those .GetValue calls will be...
2. how expensive this is:

```c#
            _initialSource = null;
            _version = textAndVersion.Version;
            _filePath = textAndVersion.FilePath;
            _loadDiagnostic = textAndVersion.LoadDiagnostic;
            _text = new RecoverableText(this, textAndVersion.Text);
            _text.GetValue(CancellationToken.None); // force access to trigger save
            return textAndVersion;
```

i don't have a good sense of what the tradeoff might be ehre.

Reply:
> how expensive all those .GetValue calls will be...

In this case, the call to `_initialSource.GetValue` is negligible because `_initialSource` is a `ConstantValueSource<TextAndVersion>`. The overall cost of the new code is essentially the cost of:

https://github.com/dotnet/roslyn/blob/bd545a6c2efc9dcca054b926589f0336b1be32bc/src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs#L156-L157

We already perform this operation as part of the `TouchDocumentAction` processing, so there are two fundamental differences with this approach:

1. Saving to temporary storage occurs _earlier_, though it still occurs exactly one time in total.
2. For the case where processing queue depth triggers a Gen 2 GC, it may cause the weak reference to go away which then needs to be loaded from temporary storage during subsequent event processing. This case is unlikely to be slower than the previous approach because this is the scenario that most often led to crashes prior to this pull request.

Reply:
> For the case where processing queue depth triggers a Gen 2 GC

Just to check, are we seeing cases where we have CompilationTrackers with multiple TouchDocumentActions queued to the same document?

Reply:
One of the heaps I have suggests yes.